forwarder.go

```diff
@@ -0,0 +1,200 @@
+				connections: 0,
+				active:      atomic.Bool{},
+			}
+			r.active.Store(true)
+			r.healthy.Store(true)
+			return r
+		}())
+	}
+	return fwd
+}
+
+// UpdateServicePool will update service pool merging the new pool routes
+// configuration with existing routes
+func (f *Forwarder) UpdateServicePool(pool ServicePool) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	// Create the match map
+	currentPoolMap := map[string]*route{}
+	for _, rte := range *f.routes {
+		currentPoolMap[rte.address] = rte
+	}
+
+	newRoutePool := make([]*route, 0)
+	for _, poolRoute := range pool.Routes() {
+		// If route exists then change parameters and inherit current connection stage
+		if fwdRoute, exists := currentPoolMap[poolRoute.Path()]; exists {
+			fwdRoute.active.Store(poolRoute.Active())
+			newRoutePool = append(newRoutePool, fwdRoute)
+			delete(currentPoolMap, poolRoute.Path())
+			continue
+		}
+		// Create new otherwise
+		newRoutePool = append(newRoutePool, func() *route {
+			r := &route{
+				address:     poolRoute.Path(),
+				healthy:     atomic.Bool{},
+				connections: 0,
+				active:      atomic.Bool{},
+			}
+			r.active.Store(poolRoute.Active())
+			r.healthy.Store(true)
+			return r
+		}())
+	}
+	// For the routes not in the new update, mark inactive for the rest of the resources free them if holding the pointer
+	for _, rte := range currentPoolMap {
+		rte.active.Store(false)
+	}
+	f.routes = &newRoutePool
+	f.logger.Info().Msgf("forwarder routes updated to: %+v from: %+v", *f.routes, pool.Routes())
+}
+
+// Attach will attach some incoming session to the pool of upstream traffic distribution
+func (f *Forwarder) Attach(ctx context.Context, in io.ReadWriteCloser) error {
+
+	errTransport := make(chan error, 2)
+	defer in.Close()
+
+	rte := f.strategy.Next()
+	if rte == nil {
+		return fmt.Errorf("no active routes available")
+	}
+
+	dest, err := net.DialTimeout("tcp", rte.address, f.dialTimeout)
+	if err != nil {
+		f.logger.Err(err).Msgf("route unreachable %s", rte.address)
+		return err
+	}
+	defer dest.Close()
+
+	// Connection increment here as we reached destination
+	atomic.AddUint32(&rte.connections, 1)
+
+	go func(w io.WriteCloser, r io.ReadCloser) {
+		defer w.Close()
+		defer r.Close()
+		_, err := io.Copy(w, r)
+		errTransport <- err
+	}(dest, in)
+
+	go func(w io.WriteCloser, r io.ReadCloser) {
+		defer w.Close()
+		defer r.Close()
+		_, err := io.Copy(w, r)
+		errTransport <- err
+	}(in, dest)
+
+	var errs []error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err = <-errTransport:
+			// If detected error, check that error has nature of a normal behavior in the system
+			// and will not affect the further behavior
+			if err != nil && !(errors.Is(err, io.EOF) || strings.Contains(err.Error(), closedNetworkConnection)) {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	// Connection decrement here as all pipes are closed by now
+	atomic.AddUint32(&rte.connections, ^uint32(0))
```

Comment:
Nit: `math.MaxUint32` is probably more readable for engineers who don't have as strong an understanding of bitwise operators.

Also - this decrement doesn't get hit if the context is cancelled. Would this be better as a defer immediately after incrementing it ?